addressmanager/src/lib.rs

```diff
@@ -0,0 +1,200 @@
+        self.banned_address_store.remove(ip).unwrap();
+    }
+
+    pub fn is_banned(&mut self, ip: Ipv6Addr) -> bool {
+        const MAX_BANNED_TIME: u64 = 24 * 60 * 60 * 1000;
+        match self.banned_address_store.get(ip).unwrap_option() {
+            Some(timestamp) => {
+                if unix_time() - timestamp > MAX_BANNED_TIME {
+                    self.unban(ip);
+                    false
+                } else {
+                    true
+                }
+            }
+            None => false,
+        }
+    }
+}
+
+fn unix_time() -> u64 {
+    SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_millis() as u64
+}
+
+mod not_banned_address_store_with_cache {
+    use std::{
+        cmp::min,
+        collections::{HashMap, HashSet},
+        net::Ipv6Addr,
+        sync::Arc,
+    };
+
+    use database::db::DB;
+    use itertools::Itertools;
+    use rand::{distributions::WeightedIndex, prelude::Distribution};
+
+    use crate::{
+        stores::not_banned_address_store::{DbNotBannedAddressesStore, NotBannedAddressesStore, NotBannedAddressesStoreReader},
+        NetAddress, MAX_ADDRESSES, MAX_CONNECTION_FAILED_COUNT,
+    };
+
+    pub struct Store {
+        db_store: DbNotBannedAddressesStore,
+        addresses: HashMap<NetAddress, u64>,
+    }
+
+    impl Store {
+        fn new(db: Arc<DB>) -> Self {
+            let db_store = DbNotBannedAddressesStore::new(db, 0);
+            let mut addresses = HashMap::new();
+            for ((ip, port), connection_failed_count) in db_store.iterator().map(|res| res.unwrap()) {
+                addresses.insert(NetAddress { ip, port }, connection_failed_count);
+            }
+
+            Self { db_store, addresses }
+        }
+
+        pub fn set(&mut self, address: NetAddress, connection_failed_count: u64) {
+            self.db_store.set(address.ip, address.port, connection_failed_count).unwrap();
+            self.addresses.insert(NetAddress::new(address.ip, address.port), connection_failed_count);
+            self.keep_limit();
+        }
+
+        fn keep_limit(&mut self) {
+            if self.addresses.len() <= MAX_ADDRESSES {
+                return;
+            }
+
+            let to_remove = self.addresses.iter().max_by(|a, b| a.1.cmp(b.1)).unwrap();
+            self.db_store.remove(to_remove.0.ip, to_remove.0.port).unwrap();
+        }
+
+        pub fn get(&self, address: NetAddress) -> u64 {
+            self.db_store.get(address.ip, address.port).unwrap()
+        }
+
+        pub fn remove(&mut self, address: NetAddress) {
+            self.db_store.remove(address.ip, address.port).unwrap()
+        }
+
+        pub fn get_all_addresses(&self) -> impl Iterator<Item = NetAddress> + '_ {
+            self.addresses.keys().copied()
+        }
+
+        pub fn get_random_addresses(&self, count: usize, exceptions: HashSet<NetAddress>) -> Vec<NetAddress> {
+            let count = min(count, self.addresses.len());
```

Comment:
I think this is safer:

```
            let addresses = self.addresses.iter().filter(|(a, _)| !exceptions.contains(a)).collect_vec();
            let mut weights =
                addresses.iter().map(|(_, &count)| 64f64.powf((MAX_CONNECTION_FAILED_COUNT + 1 - count) as f64)).collect_vec();
            let count = min(count, addresses.len());
```

having zeroed entries might make the count too large and you might try sampling from a vec of zeros which probably is an err